### PR TITLE
[SPARK-47963][CORE] Make the external Spark ecosystem can use structured logging mechanisms

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to make `the external Spark ecosystem` can use `structured logging mechanisms`.

### Why are the changes needed?
Currently, the class of MDC's key is `LogKey`, while LogKey is an `enumeration`. If we want to use structured logging mechanism in a `third-party Spark ecosystem`, we cannot currently use it (we cannot expand and `add a new LogKey enumeration value`)
So I propose changing the `LogKey` from `enumeration` to the following, 
image
as we already have a similar way of writing it in the Spark code, eg:
https://github.com/apache/spark/blob/6f01982094f6cb6e55a9edfbe06f9bfa9e3bce01/core/src/main/scala/org/apache/spark/TaskEndReason.scala#L39-L46

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Add new UT (Give an example to developers who use the LogKey mechanism in external systems related to Spark.)
- Manually test.
- Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by panbingkun** [comment]
cc @gengliangwang @LuciferYang

File: common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala [hunk 1]
```diff
@@ -144,6 +146,22 @@ trait LoggingSuiteBase
       }
   }
 
+  // An enumeration value that is not defined in `LogKey`, you can define it anywhere
+  private val externalLogKey = LogKey.VALUE
+  private val externalLog = log"${MDC(externalLogKey, "External log message.")}"
+  test("Logging with external LogKey") {
```
Comment (LuciferYang) [inline common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala:152]: For this case, the key for `externalLog.context` will be an empty string. Is this expected?

Comment (LuciferYang) [reply common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala:152]: In https://github.com/apache/spark/pull/46186/files, I try to relax the type of `MDC.key` to `Enumeration#Value`, which allows us to extend the use of other enumeration types as `MDC.key` to enhance a bit of extensibility.

**Comment by panbingkun** [comment]
Wait a moment, I'm checking

File: common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala (hunk 1 above)
Comment (LuciferYang) [reply common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala:152]: friendly ping @gengliangwang,  what's your take on this? Is `StructuredLogging` a feature that users can utilize? If so, how can we make it easier for users to extend the definition of `LogKey`?

Comment (panbingkun) [reply common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala]: @LuciferYang @gengliangwang 
I have updated a version that removes `enumeration`. When using structured logging mechanisms in external Spark related ecosystems, the following format can be used:
```
case object CUSTOME_LOG_KEY extends ILogKey
```
At present, this form is more in line with the traditional Spark style, and we can `constrain` the `class` of MDC parameter `key`

Comment (LuciferYang) [reply common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala]: Perhaps we also need to consider namespace isolation. Currently, we can define keys with the same name in multiple places, but we cannot distinguish from the `MessageWithContext` whether they are defined internally in Spark or externally.

Comment (panbingkun) [reply common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala]: If necessary in the future, I think it would be easier to implement based on the `current changes`, such as
- `ILogKey`

- `InternalLogKey` extends `ILogKey`
- `LogKey` extends `InternalLogKey`

- `ExternalLogKey` extends `ILogKey`
- `ThirdPartyLogKey` extends `ExternalLogKey`

Comment (gengliangwang) [reply common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala]: We should definitely allow external libraries to extend the log keys. But I don't have a good suggestion for this one now. Either making it as a trait or making the keys as strings are ok.

Comment (panbingkun) [reply common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala]: @gengliangwang 
What do you think of the current modification approach?
If we define it as `a string`, it feels like there is `no binding requirement` for `LogKey` at all
Perhaps we need more eyes to review the changes above?

File: common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala [hunk 2]
```diff
@@ -16,393 +16,393 @@
  */
 package org.apache.spark.internal
 
+trait ILogKey
```
Comment (gengliangwang) [inline common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:19]: rename as LogKey?

File: common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala [hunk 3]
```diff
@@ -16,393 +16,393 @@
...
-  val WAIT_SEND_TIME = Value
-  val WAIT_TIME = Value
-  val WATERMARK_CONSTRAINT = Value
-  val WEIGHTED_NUM = Value
-  val WORKER_URL = Value
-  val WRITE_AHEAD_LOG_INFO = Value
-  val XSD_PATH = Value
-
-  type LogKey = Value
+object LogKey extends ILogKey {
```
Comment (gengliangwang) [inline common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:25]: Rename as LogKeys? Also, there seems no need to extend any trait.

File: common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala (hunk 1 above)
Comment (gengliangwang) [reply common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala]: Yeah creating a trait seems better. I have some preference in the naming. See my other comments.

File: common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala (hunk 3 above)
Comment (panbingkun) [reply common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:25]: @gengliangwang 
There is `a small` question, I actually think this name is `not good`, but I made a few compromises. If we change this name, it will bring import changes to all classes that use it. This may involve many files, can we accept it? eg:
```
import org.apache.spark.internal.LogKey.{CODEC_LEVEL, CODEC_NAME, CONFIG, PATH}

=>

import org.apache.spark.internal.LogKeys.{CODEC_LEVEL, CODEC_NAME, CONFIG, PATH}

```

image

About `200+`, it sounds like it's okay too, isn't it?

Comment (gengliangwang) [reply common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:25]: Yes, it is totally fine. I also tried it on my local. Looks good.

Comment (panbingkun) [reply common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:25]: `Also, there seems no need to extend any trait.`
Yeah, You are right, this is an intermediate process that I forgot to remove from my changes.

Comment (panbingkun) [reply common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:25]: > Yes, it is totally fine. I also tried it on my local. Looks good.

Okay, let me handle it.

File: common/utils/src/main/scala/org/apache/spark/internal/README.md [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 ## LogKey
 
-LogKeys serve as identifiers for mapped diagnostic contexts (MDC) within logs. Follow these guidelines when adding new LogKeys:
+LogKey serve as identifiers for mapped diagnostic contexts (MDC) within logs. Follow these guidelines when adding new LogKey:
```
Comment (panbingkun) [inline common/utils/src/main/scala/org/apache/spark/internal/README.md:5]: I'm not sure if it's more accurate to call it `LogKeys` or `LogKey`?

File: common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala [hunk 5]
```diff
@@ -144,6 +146,21 @@ trait LoggingSuiteBase
       }
   }
 
+  private val externalSystemCustomLog =
```
Comment (panbingkun) [inline common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala:149]: Add a UT to test how external third-party can use `LogKey`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeq.scala [hunk 6]
```diff
@@ -21,12 +21,12 @@ import org.json4s.{Formats, NoTypeHints}
...
 
 import org.apache.spark.internal.{Logging, MDC}
-import org.apache.spark.internal.LogKey.{CONFIG, DEFAULT_VALUE, NEW_VALUE, OLD_VALUE, TIP}
+import org.apache.spark.internal.LogKeys.{CONFIG, DEFAULT_VALUE, NEW_VALUE, OLD_VALUE, TIP}
 import org.apache.spark.io.CompressionCodec
 import org.apache.spark.sql.RuntimeConfig
 import org.apache.spark.sql.connector.read.streaming.{Offset => OffsetV2, SparkDataStream}
 import org.apache.spark.sql.execution.streaming.state.{FlatMapGroupsWithStateExecHelper, StreamingAggregationStateManager, SymmetricHashJoinStateManager}
-import org.apache.spark.sql.internal.SQLConf.{FLATMAPGROUPSWITHSTATE_STATE_FORMAT_VERSION, _}
+import org.apache.spark.sql.internal.SQLConf._
```
Comment (panbingkun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeq.scala:29]: This is not related to the renaming of `LogKey -> LogKeys`, but it is almost more reasonable to change it to this way.

File: common/utils/src/main/scala/org/apache/spark/internal/README.md (hunk 4 above)
Comment (gengliangwang) [inline common/utils/src/main/scala/org/apache/spark/internal/README.md:5]:
```suggestion
`LogKey`s serve as identifiers for mapped diagnostic contexts (MDC) within logs. Follow these guidelines when adding a new LogKey:
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 7]
```diff
@@ -20,15 +20,15 @@ package org.apache.spark.sql.catalyst.optimizer
...
+import org.apache.spark.internal.LogKeys._
 import org.apache.spark.internal.MDC
 import org.apache.spark.sql.catalyst.SQLConfHelper
 import org.apache.spark.sql.catalyst.analysis._
 import org.apache.spark.sql.catalyst.catalog.{InMemoryCatalog, SessionCatalog}
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.expressions.aggregate._
 import org.apache.spark.sql.catalyst.plans._
-import org.apache.spark.sql.catalyst.plans.logical.{RepartitionOperation, _}
+import org.apache.spark.sql.catalyst.plans.logical._
```
Comment (panbingkun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:31]: This is not related to the renaming of LogKey -> LogKeys, but it is almost more reasonable to change it to this way.

File: common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala [hunk 8]
```diff
@@ -16,399 +16,399 @@
  */
 package org.apache.spark.internal
 
+trait LogKey
```
Comment (LuciferYang) [inline common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:19]: Relevant explanatory comments should be added for `LogKey`

File: common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala [hunk 9]
```diff
@@ -16,399 +16,399 @@
...
+  case object TEMP_FILE extends LogKey
+  case object TEMP_PATH extends LogKey
+  case object TEST_SIZE extends LogKey
+  case object THREAD extends LogKey
+  case object THREAD_NAME extends LogKey
+  case object TID extends LogKey
+  case object TIME extends LogKey
+  case object TIMEOUT extends LogKey
+  case object TIMER extends LogKey
+  case object TIMER_LABEL extends LogKey
```
Comment (LuciferYang) [inline common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:371]: `TIMER_LABEL` and `TIME_UNITS` are not being used, should we temporarily clean them up?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 10]
```diff
@@ -33,8 +33,8 @@ import org.apache.spark.InternalAccumulator.{input, shuffleRead}
 import org.apache.spark.TaskState.TaskState
 import org.apache.spark.errors.SparkCoreErrors
 import org.apache.spark.executor.ExecutorMetrics
-import org.apache.spark.internal.{config, Logging, LogKey, MDC}
-import org.apache.spark.internal.LogKey.{REASON, TASK_SET_NAME, TASK_STATE, TID}
+import org.apache.spark.internal.{config, Logging, LogKeys, MDC}
+import org.apache.spark.internal.LogKeys.{REASON, TASK_SET_NAME, TASK_STATE, TID}
```
Comment (panbingkun) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:37]: merge to `import org.apache.spark.internal.LogKeys`

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala [hunk 11]
```diff
@@ -33,8 +33,8 @@ import org.apache.spark.deploy.k8s.Config._
 import org.apache.spark.deploy.k8s.Constants._
 import org.apache.spark.deploy.k8s.KubernetesConf
 import org.apache.spark.deploy.k8s.KubernetesUtils.addOwnerReference
-import org.apache.spark.internal.{Logging, LogKey, MDC}
-import org.apache.spark.internal.LogKey.{COUNT, EXECUTOR_IDS, TIMEOUT}
+import org.apache.spark.internal.{Logging, LogKeys, MDC}
+import org.apache.spark.internal.LogKeys.{COUNT, EXECUTOR_IDS, TIMEOUT}
```
Comment (panbingkun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:37]: merge to `import org.apache.spark.internal.LogKeys`

File: common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala (hunk 9 above)
Comment (panbingkun) [reply common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:371]: Let me remove it.
When all pr for the structured logs is completed, I will also submit a PR to `improve` the names of this class, such as `NUM-FILES` and `WEIGHTED-NUM`. Should `NUM` be used as the prefix or suffix? We need to unify. @gengliangwang @LuciferYang

Comment (gengliangwang) [reply common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala]: prefix sounds good

File: common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala (hunk 1 above)
Comment (LuciferYang) [reply common/utils/src/test/scala/org/apache/spark/util/StructuredLoggingSuite.scala]: > If necessary in the future, I think it would be easier to implement based on the `current changes`, such as
> 
> * `ILogKey`
> * `InternalLogKey` extends `ILogKey`
> * `LogKey` extends `InternalLogKey`
> * `ExternalLogKey` extends `ILogKey`
> * `ThirdPartyLogKey` extends `ExternalLogKey`

Yes, we can further expand after this PR is completed. 

Personally, I think we can add a String field to the MDC to identify the namespace. For example, the namespace of Spark's internal MDC can always be `SPARK`, and the third-party Spark ecosystem can customize their own namespace. This can make the key of `MessageWithContext` become similar to `SPARK.mdcKey`, `CUSTOM_NAMESPACE.mdcKey`. 

Perhaps this is not a perfect method, we can further discuss it after this PR is completed.

File: common/utils/src/test/scala/org/apache/spark/util/LogKeySuite.scala [hunk 12]
```diff
@@ -57,20 +57,20 @@ class LogKeySuite
...
     if (originalKeys != sortedKeys) {
       val logKeyFile = logKeyFilePath.toFile
-      logInfo(s"Regenerating LogKey file $logKeyFile")
+      logInfo(s"Regenerating the file $logKeyFile")
       val originalContents = FileUtils.readLines(logKeyFile, StandardCharsets.UTF_8)
       val sortedContents = new JList[String]()
       var firstMatch = false
       originalContents.asScala.foreach { line =>
-        if (line.trim.startsWith("val ") && line.trim.endsWith(" = Value")) {
+        if (line.trim.startsWith("case object ") && line.trim.endsWith(" extends LogKey")) {
```
Comment (gengliangwang) [inline common/utils/src/test/scala/org/apache/spark/util/LogKeySuite.scala:70]: Not directly related to this PR, but we need to think about what if there are code comments around the LogKey.

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
@panbingkun Sorry, I have to merge https://github.com/apache/spark/pull/46192 since it was created before this one. Please resolve the conflicts and I will merge this one.  Thanks as always.

**Comment by panbingkun** [comment]
> @panbingkun Sorry, I have to merge #46192 since it was created before this one. Please resolve the conflicts and I will merge this one. Thanks as always.

Don't worry, I have solved it, 😄
Let's run GA first.

**Review by LuciferYang — APPROVED** [review]

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master
